### PR TITLE
daemons: changing our custom daemon CLI parsing with CLI11

### PR DESCRIPTION
**Pull request by alaaeddineelamri**
### Description

This PR aims to standardize daemon argument parsing using a third party library called CLI11.
Documentation can be found [here](https://github.com/CLIUtils/CLI11) and [also here](https://cliutils.github.io/CLI11/book/).

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

##### General
- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Separate commit for this PR in the CHANGELOG.md, PR number referenced is same
- [x] Commit descriptions are understandable and well formatted
- ~~[ ] If backport: add original PR number and target branch at top of this file: **Backport of PR#000 to bareos-2x**~~

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR
- [x] `bareos-check-sources --since-merge` does not report any problems
- [x] `git status` should not report modifications in the source tree after building and testing

##### Tests

- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: core/src/stored/bextract.cc [hunk 1]
```diff
@@ -131,94 +103,144 @@ int main(int argc, char* argv[])
...
+  std::string archive_device_name;
+  bextract_app
+      .add_option("bareos-archive-device-name", archive_device_name,
+                  "Specify input archive names.")
+      ->required()
+      ->type_name(" ");
+
+  std::string directory_to_store_files;
+  bextract_app
+      .add_option("directory-to-store-files", directory_to_store_files,
```
Comment (joergsteffens) [inline core/src/stored/bextract.cc:232]:
```suggestion
      .add_option("target-directory", directory_to_store_files,
```

File: core/src/stored/bextract.cc [hunk 2]
```diff
@@ -382,15 +404,18 @@ static void ClosePreviousStream(void)
...
+                      DirectorResource* director)
 {
   struct stat statp;
   uint32_t decompress_buf_size;
 
   EnableBackupPrivileges(NULL, 1);
 
   dcr = new DeviceControlRecord;
-  jcr = SetupJcr("bextract", devname, bsr, director, dcr, VolumeName,
+  jcr = SetupJcr("bextract", devname, bsr, director, dcr, VolumeName.data(),
```
Comment (joergsteffens) [inline core/src/stored/bextract.cc:418]: `data()`? Not `c_str()`?

File: core/src/stored/bls.cc [hunk 3]
```diff
@@ -130,99 +94,131 @@ int main(int argc, char* argv[])
...
+      .add_flag(
+          "-v,--verbose", [](bool val) { verbose++; }, "Verbose user messages.")
+      ->trigger_on_parse();
 
-      case 'V': /* Volume name */
-        VolumeName = optarg;
-        break;
+  std::string VolumeName;
+  bls_app
+      .add_option("-V,--volume-names", VolumeName,
```
Comment (joergsteffens) [inline core/src/stored/bls.cc:211]:
```suggestion
      .add_option("-V,--volumes", VolumeName,
```

File: core/src/stored/bls.cc [hunk 4]
```diff
@@ -130,99 +94,131 @@ int main(int argc, char* argv[])
...
+                  "Volume names (separated by |)")
+      ->type_name("<vol1|vol2|...>");
 
-      case '?':
-      default:
-        usage();
+  std::vector<std::string> archive_names;
+  bls_app
+      .add_option("archive-names", archive_names, "Specify input archive names")
```
Comment (joergsteffens) [inline core/src/stored/bls.cc:217]: You should use the opportunity to better specify this parameter (identical in all the Bareos Volume Utilities), as this has been a source of trouble for a lot of people. Maybe:
```
                  "Specify the input device name (either as name of a Bareos Storage Daemon Device resource or identical to the Archive Device in a Bareos Storage Daemon Device resource .")
```

File: core/src/stored/bls.cc [hunk 5]
```diff
@@ -130,99 +94,131 @@ int main(int argc, char* argv[])
...
+  std::string VolumeName;
+  bls_app
+      .add_option("-V,--volume-names", VolumeName,
+                  "Volume names (separated by |)")
+      ->type_name("<vol1|vol2|...>");
 
-      case '?':
-      default:
-        usage();
+  std::vector<std::string> archive_names;
```
Comment (joergsteffens) [inline core/src/stored/bls.cc:215]: I see this more as devices than as archives (YMMV).

File: core/src/stored/bscan.cc [hunk 6]
```diff
@@ -197,116 +144,142 @@ int main(int argc, char* argv[])
...
-    Pmsg0(0, _("Wrong number of arguments: \n"));
-    usage();
-  }
+  CLI::App bscan_app;
+  InitCLIApp(bscan_app, "The Bareos Database Scan tool.", 2001);
+
+  std::string backend_directory = PATH_BAREOS_BACKENDDIR;
+  bscan_app
+      .add_option("-a,--backend-directory", backend_directory,
+                  "Set backend directory.")
```
Comment (joergsteffens) [inline core/src/stored/bscan.cc:153]:
```suggestion
                  "Specify a directory from where Bareos backends can be loaded..")
```

File: core/src/stored/bscan.cc [hunk 7]
```diff
@@ -197,116 +144,142 @@ int main(int argc, char* argv[])
...
+      ->type_name("<directory>");
+
+  std::string ignored_B;
+  bscan_app
+      .add_option("-B,--db-driver", ignored_B,
```
Comment (joergsteffens) [inline core/src/stored/bscan.cc:158]:
```suggestion
      .add_option("-B,--dbdriver", ignored_B,
```

File: core/src/stored/bscan.cc [hunk 8]
```diff
@@ -197,116 +144,142 @@ int main(int argc, char* argv[])
...
+
+  bscan_app.add_flag("-S,--show-progress", showProgress,
+                     "Show scan progress periodically.");
+
+  bscan_app.add_flag("-s,--update-db", update_db,
+                     "Synchronize or store in database.");
+
+  std::string VolumeName;
+  bscan_app
+      .add_option("-V,--volume-name", VolumeName,
```
Comment (joergsteffens) [inline core/src/stored/bscan.cc:248]:
```suggestion
      .add_option("-V,--volumes", VolumeName,
```

File: core/src/stored/btape.cc [hunk 9]
```diff
@@ -205,99 +199,122 @@ int main(int margc, char* margv[])
...
+      ->check(CLI::NonNegativeNumber)
+      ->type_name("<level>");
+
+  btape_app.add_flag("--dt,--debug-timestamp", dbg_timestamp,
+                     "Print timestamp in debug output.");
+
+  btape_app.add_flag("-p,--proceed-io", forge_on,
+                     "Proceed inspite of IO errors");
+
+  btape_app.add_flag("-s,--turnoff-signals", turnoff_signals,
```
Comment (joergsteffens) [inline core/src/stored/btape.cc:278]:
```suggestion
  btape_app.add_flag("-s{false},--no-signals{false}", signals,
```
This should work also is is more intuitive (initializing signals=true, of course).

File: core/src/stored/stored.cc [hunk 10]
```diff
@@ -155,104 +120,98 @@ int main(int argc, char* argv[])
...
   }
 
-#if HAVE_WIN32
-  std::string allowed_parameters("c:d:fg:imstu:vx:z:?");
+  CLI::App sd_app;
+  InitCLIApp(sd_app, "The Bareos Storage Daemon.", 2000);
+
+  sd_app
+      .add_option(
+          "configuration,-c,--config",
```
Comment (joergsteffens) [inline core/src/stored/stored.cc:128]:
```suggestion
          "-c,--config",
```

File: core/src/stored/stored.cc [hunk 11]
```diff
@@ -155,104 +120,98 @@ int main(int argc, char* argv[])
...
+            if (configfile != nullptr) { free(configfile); }
+            configfile = strdup(val.front().c_str());
+            return true;
+          },
+          "Use <path> as configuration file or directory.")
+      ->check(CLI::ExistingPath)
+      ->type_name("<path>");
+
+  sd_app
+      .add_option("-d,--debug-level", debug_level,
```
Comment (joergsteffens) [inline core/src/stored/stored.cc:139]:
```suggestion
      .add_option("-d,--debug", debug_level,
```

File: core/src/stored/stored.cc [hunk 12]
```diff
@@ -155,104 +120,98 @@ int main(int argc, char* argv[])
...
-        if (*optarg == 's') {
-          export_config_schema = true;
-        } else if (*optarg == 'c') {
-          export_config = true;
-        } else {
-          usage();
-        }
-        break;
+  bool no_signals = false;
+  sd_app.add_flag("-s,--turnoff-signals", no_signals,
```
Comment (joergsteffens) [inline core/src/stored/stored.cc:175]:
```suggestion
  sd_app.add_flag("-s{false},--no-signals{false}", signals,
```

**Review by joergsteffens — COMMENTED** [review]
I'm not through all files, howver wanted to provide you the part of the review already done.

File: core/manpages/bareos-fd.8 [hunk 13]
```diff
@@ -24,49 +24,58 @@ backup'd/verified/restored. Network transfer can optionally be
...
+.BI \-u,--user\  user
 Set the username/uid to run as.
 .TP
-.BI \-v
+.BI \-v,--verbose
 Set verbose mode.
 .TP
-.B \-?
+.BI \--xc,--export-config\ [<resource> [<name>]]
+Print all or specific configuration resources and exit.
```
Comment (joergsteffens) [inline core/manpages/bareos-fd.8:70]:
```suggestion
Print configuration and exit.
```

File: core/manpages/bareos-sd.8 [hunk 14]
```diff
@@ -22,43 +22,52 @@ network backup system and a tape drive/autochanger or filesystem where
...
+.BI \-u,--user\  user
 Set the username/uid to run as.
 .TP
-.BI \-v
+.BI \-v,--verbose
 Set verbose mode.
 .TP
-.B \-?
+.BI \--xc,--export-config\ [<resource> [<name>]]
+Print all or specific configuration resources and exit.
```
Comment (joergsteffens) [inline core/manpages/bareos-sd.8:62]:
```suggestion
.BI \--xc,--export-config
Print configuration and exit.
```

File: core/src/stored/bcopy.cc [hunk 15]
```diff
@@ -62,123 +63,124 @@ static Device* out_dev = NULL;
...
-        break;
+  bcopy_app
+      .add_flag(
+          "-v,--verbose", [](bool val) { verbose++; }, "Verbose user messages.")
+      ->trigger_on_parse();
 
-      case 'w':
-        wd = optarg;
-        break;
+  std::string work_dir = "/tmp";
```
[the remark this answers is not in the document]
Comment (alaaeddineelamri) [reply core/src/stored/bcopy.cc:158]: what do you mean? the default value `/tmp` or the variable `work_dir`? 
If it's the default value, `/tmp` value was already the default. In that case we add a different default value for windows, or just keep it empty.

File: core/src/stored/btape.cc (hunk 9 above)
Comment (alaaeddineelamri) [reply core/src/stored/btape.cc:278]: You are right, it is possible and I tried it. Same goes for `!-s,!--no-signals`.
But if you use any of them, and you show the help, it looks like this:
```
-s{false},--no-signals{false}
```
which is in my opinion confusing

File: core/src/dird/dird.cc [hunk 16]
```diff
@@ -242,121 +200,123 @@ int main(int argc, char* argv[])
...
-      case '?':
-      default:
-        usage();
-        break;
-    }
-  }
-  argc -= optind;
-  argv += optind;
+  dir_app.add_flag("-t,--test", test_config,
+                   "Test - read configuration and exit.");
```
[the remark this answers is not in the document]
Comment (alaaeddineelamri) [reply core/src/dird/dird.cc:264]: `-f` and `-t` are quite extensively used together to do checks in scripts. Making them exclude each other breaks a lot of tests. What makes more sense I think is to make only `--pidfile` exclude `-f` and `-t`, and let `-f` and `-t` coexist.

File: core/src/stored/bextract.cc [hunk 17]
```diff
@@ -131,94 +103,144 @@ int main(int argc, char* argv[])
...
+      ->trigger_on_parse();
+
+  std::string VolumeName;
+  bextract_app
+      .add_option("-V,--volume-names", VolumeName,
+                  "Volume names (separated by |).")
+      ->type_name("<vol1|vol2|...>");
+
+
+  std::string work_dir = "/tmp";
```
[the remark this answers is not in the document]
Comment (alaaeddineelamri) [reply core/src/stored/bextract.cc:217]: turns out even in the previous code, it was not used. It is given the value `/tmp` but the variable is not used at all. 
I should then just delete the option, and just keep the `working_directory` as is, or just give it the current path as you suggested.

File: core/src/dird/dird.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (joergsteffens) [reply core/src/dird/dird.cc:264]: You're right. I misread the code.

File: third-party/README.md [hunk 18]
```diff
@@ -0,0 +1,14 @@
+# Bareos third party libraries
+
+This is the "home" of external libraries that are used in Bareos
+
+### Adding third party libraries
+
+Each Library will be put into its own subdirectory with `git subtree` and will have its own CMake include-file (if it is not already made available by the provider of the library), that is then loaded in CMakeLists.txt
+
```
Comment (joergsteffens) [inline third-party/README.md:8]: We need information about how to handle the licenses of the third-party libraries.

**Review by joergsteffens — CHANGES_REQUESTED** [review]

**Review by joergsteffens — APPROVED** [review]
Good job!

**Pull request merged**